src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlObjectSerializerReadContext.cs

```diff
@@ -411,27 +425,31 @@ public void ReplaceDeserializedObject(string id, object oldObj, object newObj)
                 // These XmlObjectSerializer implementations do not currently support fix-ups. Hence we
                 // throw in such cases to allow us add fix-up support in the future if we need to.
                 if (DeserializedObjects.IsObjectReferenced(id))
+                {
+                    // https://github.com/dotnet/runtime/issues/41465 - oldObj or newObj may be null below - suppress compiler error by asserting non-null
+                    Debug.Assert(oldObj != null && newObj != null);
                     throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(XmlObjectSerializer.CreateSerializationException(SR.Format(SR.FactoryObjectContainsSelfReference, DataContract.GetClrTypeFullName(oldObj.GetType()), DataContract.GetClrTypeFullName(newObj.GetType()), id)));
+                }
                 DeserializedObjects.Remove(id);
                 DeserializedObjects.Add(id, newObj);
             }
         }
 
 #if USE_REFEMIT
-        public object GetExistingObject(string id, Type type, string name, string ns)
+        public object GetExistingObject(string id, Type? type, string? name, string? ns)
 #else
-        internal object GetExistingObject(string id, Type type, string name, string ns)
+        internal object GetExistingObject(string id, Type? type, string? name, string? ns)
 #endif
         {
-            object retObj = DeserializedObjects.GetObject(id);
+            object? retObj = DeserializedObjects.GetObject(id);
             if (retObj == null)
                 throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(XmlObjectSerializer.CreateSerializationException(SR.Format(SR.DeserializedObjectWithIdNotFound, id)));
             return retObj;
         }
 
         private object GetExistingObjectOrExtensionData(string id)
         {
-            object retObj = DeserializedObjects.GetObject(id);
+            object? retObj = DeserializedObjects.GetObject(id);
```

Comment:
For `public object GetRealObject(IObjectReference obj, string id)` the comment suggests it might return null so I'd expect return value to be nullable

Reply:
Fixed.